src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Receivers.kt

```diff
@@ -150,11 +152,87 @@ class MarkdownTaskReceiver(
     private val arguments = listOf("projectRootAbsPath", "workingDirRelPath",
             "inputFileName", "outputFileName", "mainFont")
     private val COMPILE_COMMAND_KEY = "pandoc.compile.command"
+    // TODO: fixedThreadPool? if so, how many threads?
+    private val executor = Executors.newFixedThreadPool(4)
+    private val currentTasks = ConcurrentHashMap<String, Future<Boolean>>()
 
     override fun processMessage(message: PubsubMessage): Boolean {
-        val request = CompileRequest.parseFrom(message.data)
+        val request = Request.parseFrom(message.data)
+
+        when (request.typeCase) {
+            Request.TypeCase.COMPILE -> return processCompile(request.compile)
+            Request.TypeCase.CANCEL  -> return processCancel(request.cancel)
+            else -> {
+                LOGGER.error("Request type is not set")
+            }
+        }
+
+        return true
+    }
+
+    private fun processCancel(request: CancelRequestProto): Boolean {
+        LOGGER.debug("Canceling the task with id = {}", request.taskId)
+        var status = CompilerFleet.Cancel.Status.FAILED // init value
+        val elapsedTime = measureTimeMillis {
+            status = cancelTask(request)
+        }
+
+        val data = CompilerFleet.Cancel.newBuilder()
+                .setTaskId(request.taskId)
+                .setStatus(status)
+                .setCpuTime(elapsedTime.toInt())
+                .build()
+                .toByteString()
+        val onPublishFailureCallback = {
+            LOGGER.info("Publish cancel task ${request.taskId} response failed with code ${StatusCode.FAILURE}")
+        }
+
+        resultPublisher.publish(data, onPublishFailureCallback)
+        return true
+    }
+
+    private fun cancelTask(request: CancelRequestProto): CompilerFleet.Cancel.Status {
+        return if (!currentTasks.contains(request.taskId)) {
+            CompilerFleet.Cancel.Status.NOT_FOUND
+        } else {
+            val result = currentTasks[request.taskId]?.cancel(true)!!
+            currentTasks.remove(request.taskId)
+            // will canceling the thread solve the problem?
```

Comment:
Cancelling will not solve anything. Besides that [it doesn't work in general](http://blog.tremblay.pro/2017/08/supply-async.html), it won't work in our case either because most of the time we're running docker container and/or waiting for response from texbe (which in turn is also running docker container). So we need to remember the current long-running stage and either run `docker rm` or call texbe with the same cancellation request. 

And I really recommend doing it in a separate pull request. 

Also, I don't think that we want to accept task if there is one which is already running, unless it is cancellation of a running task. 

Here is a guide on [using PubSub with long-running tasks](https://cloud.google.com/solutions/using-cloud-pub-sub-long-running-tasks). TL;DR: use pull workers (we do), do not acknowledge until you complete execution (that's what we do as well), setup acknowledge deadline appropriately, run blocking code in each puller (apparently we do, probably unintentionally). Long story short, if we had no cancellations, we would be good to go. However, we _want_ to get new tasks when we have one task running and inspect if a new task is cancellation of the one which is running, and this will require some changes in the code.

Reply:
This was actually the most doubtful moment, which I wanted to discuss.
Ok, I will fix this in the separate pull request.